rustls-libssl/tests/config.c

```diff
@@ -0,0 +1,67 @@
+/**
+ * Exercises openssl functions like `SSL_CONF_cmd_value_type`
+ */
+
+#include <assert.h>
+#include <stdio.h>
+
+#include <openssl/ssl.h>
+
+#define CUSTOM_PREFIX "Rustls-"
+
+static const int conf_flags[] = {SSL_CONF_FLAG_SERVER, SSL_CONF_FLAG_CLIENT,
+                                 SSL_CONF_FLAG_CERTIFICATE};
+
+#define NUM_FLAGS (sizeof(conf_flags) / sizeof(conf_flags[0]))
+
+static const char *supported_cmds[] = {
+    "-min_protocol", CUSTOM_PREFIX "min_protocol",
+    "MinProtocol",   CUSTOM_PREFIX "MinProtocol",
+
+    "-max_protocol", CUSTOM_PREFIX "max_protocol",
+    "MaxProtocol",   CUSTOM_PREFIX "MaxProtocol",
+};
+
+#define NUM_CMDS (sizeof(supported_cmds) / sizeof(supported_cmds[0]))
+
+void test_cmds(int base_flags, const char *prefix) {
+  SSL_CONF_CTX *cctx = SSL_CONF_CTX_new();
+  assert(cctx != NULL);
+  assert(SSL_CONF_CTX_set1_prefix(cctx, prefix));
+
+  int flags = base_flags;
+  for (unsigned long i = 0; i <= NUM_FLAGS; i++) {
+    unsigned int new_flags = SSL_CONF_CTX_set_flags(cctx, flags);
+    printf("cctx flags = %u\n", new_flags);
+
+    for (unsigned long j = 0; j < NUM_CMDS; j++) {
+      const char *cmd = supported_cmds[j];
+      int value = SSL_CONF_cmd_value_type(cctx, cmd);
+      printf("\tcmd %s has value type %d\n", cmd, value);
+    }
+
+    if (i < NUM_FLAGS) {
+      flags |= conf_flags[i];
+    }
+  }
+
+  assert(SSL_CONF_CTX_finish(cctx));
+  SSL_CONF_CTX_free(cctx);
+}
+
+int main(void) {
+  printf("no base flags, default prefix:\n");
+  test_cmds(0, "");
+  printf("no base flags, custom prefix:\n");
+  test_cmds(0, CUSTOM_PREFIX);
+
+  printf("CMDLINE base flags, default prefix:\n");
+  test_cmds(SSL_CONF_FLAG_CMDLINE, "");
+  printf("CMDLINE base flags,custom prefix:\n");
+  test_cmds(SSL_CONF_FLAG_CMDLINE, CUSTOM_PREFIX);
+
+  printf("FILE base flags, default prefix:\n");
+  test_cmds(SSL_CONF_FLAG_FILE, "");
+  printf("FILE base flags, custom prefix:\n");
+  test_cmds(SSL_CONF_FLAG_FILE, CUSTOM_PREFIX);
+}
```

Comment:
WDYT about calling `SSL_CONF_cmd` in this test, with a variety of supported, defined-by-openssl-but-unsupported-by-us, and fictional `cmd`s?

Reply:
@ctz I'll play with this a little bit. I think passing through commands that are unsupported by our impl but supported by openssl will create output divergences that would fail the test, but passing fictional cmds should be workable.

Reply:
I pushed some updates to the `config.c` tests. This was useful, it shook out a couple divergences:

1. I was using `try_str!` in the `_SSL_CONF_cmd` entry-point on the `value: *mut c_char` arg but we may want to support commands taking `NULL` as an argument (switch-style commands), and it meant we returned `0` (error) instead of `-2` (unknown cmd) for an invocation like `SSL_CONF_cmd(cctx, "does-not-exist", NULL);`. I've reworked to pass on an `Option<&str>` from `string_from_cstring` for the value and individual commands can choose how to handle that.
2. Apparently the `MinProtocol`/`MaxProtocol` cmds return `0` for valid protocol string values if no `CTX` or `SSL` is set :shrug: I've updated my impl to match this.